### ISSUE TITLE
Failing tests with python3.8

### ISSUE THREAD

**Issue by sanjayankur31** [pauldmccarthy/fslpy#5]
Hello,

I was updating the Fedora package to version 2.5.0 but these two tests fail:

```
=================================== FAILURES ===================================
__________________________________ test_warn ___________________________________
    def test_warn():
        deprecated.resetWarningCache()
        with warnings.catch_warnings(record=True) as w:
            warnings.simplefilter('always')
            emit_warning()
            assert len(w) == 1
>           _check_warning(w[0], 'blag', WARNING_LINE_NUMBER)
tests/test_deprecated.py:39: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
w = <warnings.WarningMessage object at 0xaae65328>, name = 'blag', lineno = 13
    def _check_warning(w, name, lineno):
        assert issubclass(w.category, DeprecationWarning)
        assert '{} is deprecated'.format(name) in str(w.message)
        assert 'test_deprecated.py' in str(w.filename)
>       assert w.lineno == lineno
E       assert 12 == 13
E         -12
E         +13
tests/test_deprecated.py:31: AssertionError
_______________________________ test_deprecated ________________________________
    def test_deprecated():
        deprecated.resetWarningCache()
        with warnings.catch_warnings(record=True) as w:
            warnings.simplefilter('always')
            call_dep_func()
            assert len(w) == 1
>           _check_warning(w[0], 'depfunc', DEPRECATED_LINE_NUMBER)
tests/test_deprecated.py:54: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
w = <warnings.WarningMessage object at 0xaada78b0>, name = 'depfunc'
lineno = 23
    def _check_warning(w, name, lineno):
        assert issubclass(w.category, DeprecationWarning)
        assert '{} is deprecated'.format(name) in str(w.message)
        assert 'test_deprecated.py' in str(w.filename)
>       assert w.lineno == lineno
E       assert 22 == 23
E         -22
E         +23
tests/test_deprecated.py:31: AssertionError
=============================== warnings summary ===============================
/usr/lib/python3.8/site-packages/_pytest/mark/structures.py:330
  /usr/lib/python3.8/site-packages/_pytest/mark/structures.py:330: PytestUnknownMarkWarning: Unknown pytest.mark.piltest - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/latest/mark.html
    warnings.warn(
/usr/lib/python3.8/site-packages/_pytest/mark/structures.py:330
  /usr/lib/python3.8/site-packages/_pytest/mark/structures.py:330: PytestUnknownMarkWarning: Unknown pytest.mark.meshtest - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/latest/mark.html
    warnings.warn(
/usr/lib/python3.8/site-packages/_pytest/mark/structures.py:330
/usr/lib/python3.8/site-packages/_pytest/mark/structures.py:330
  /usr/lib/python3.8/site-packages/_pytest/mark/structures.py:330: PytestUnknownMarkWarning: Unknown pytest.mark.longtest - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/latest/mark.html
    warnings.warn(
/usr/lib/python3.8/site-packages/_pytest/mark/structures.py:330
  /usr/lib/python3.8/site-packages/_pytest/mark/structures.py:330: PytestUnknownMarkWarning: Unknown pytest.mark.igziptest - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/latest/mark.html
    warnings.warn(
fsl/utils/transform.py:17
  /builddir/build/BUILD/fslpy-2.5.0/fsl/utils/transform.py:17: DeprecationWarning: fsl.utils.transform is deprecated from version 2.4.0 and will be removed in 3.0.0. Use the fsl.transform module instead
    deprecated.warn('fsl.utils.transform',
/usr/lib/python3.8/site-packages/_pytest/mark/structures.py:330
  /usr/lib/python3.8/site-packages/_pytest/mark/structures.py:330: PytestUnknownMarkWarning: Unknown pytest.mark.unixtest - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/latest/mark.html
    warnings.warn(
/usr/lib/python3.8/site-packages/_pytest/mark/structures.py:330
  /usr/lib/python3.8/site-packages/_pytest/mark/structures.py:330: PytestUnknownMarkWarning: Unknown pytest.mark.noroottest - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/latest/mark.html
    warnings.warn(
tests/test_freesurfer.py::test_FreesurferMesh_create
tests/test_freesurfer.py::test_FreesurferMesh_create_loadall
tests/test_freesurfer.py::test_loadVertices
tests/test_freesurfer.py::test_loadVertexData_morph
tests/test_freesurfer.py::test_loadVertexData_label
tests/test_freesurfer.py::test_loadVertexData_mgh
tests/test_freesurfer.py::test_loadVertexData_annot
  /usr/lib/python3.8/site-packages/nibabel/freesurfer/io.py:65: UserWarning: Unknown extension code.
    warnings.warn("Unknown extension code.")
tests/test_freesurfer.py::test_FreesurferMesh_create
tests/test_freesurfer.py::test_FreesurferMesh_create_loadall
tests/test_freesurfer.py::test_loadVertices
tests/test_freesurfer.py::test_loadVertexData_morph
tests/test_freesurfer.py::test_loadVertexData_label
tests/test_freesurfer.py::test_loadVertexData_mgh
tests/test_freesurfer.py::test_loadVertexData_annot
  /usr/lib/python3.8/site-packages/nibabel/freesurfer/io.py:192: UserWarning: No volume information contained in the file
    warnings.warn('No volume information contained in the file')
tests/test_gifti.py::test_loadGiftiVertexData
  /usr/lib/python3.8/site-packages/nibabel/gifti/parse_gifti_fast.py:228: UserWarning: Actual # of data arrays does not match # expected: 4 != 3.
    warnings.warn("Actual # of data arrays does not match "
tests/test_run.py::test_run_submit
  /builddir/build/BUILD/fslpy-2.5.0/fsl/utils/run.py:194: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3,and in 3.9 it will stop working
    if submit is not None and not isinstance(submit, collections.Mapping):
-- Docs: https://docs.pytest.org/en/latest/warnings.html
----------- coverage: platform linux, python 3.8.0-beta-3 ------------
```

Full build log and root log with package versions attached too:
[buildlog.txt](https://github.com/pauldmccarthy/fslpy/files/3562516/buildlog.txt)
[rootlog.txt](https://github.com/pauldmccarthy/fslpy/files/3562517/rootlog.txt)

**Comment by sanjayankur31** [comment]
Another failure when disabling these two:

```
=================================== FAILURES ===================================
_________________________ test_convertDeformationType __________________________
    def test_convertDeformationType():
    
        relfield = _random_field()
        coords   = _field_coords(relfield)
        absfield = nonlinear.DeformationField(
            relfield.data + coords,
            src=relfield.src,
            xform=relfield.voxToWorldMat)
    
        gotconvrel1 = nonlinear.convertDeformationType(relfield)
        gotconvabs1 = nonlinear.convertDeformationType(absfield)
        gotconvrel2 = nonlinear.convertDeformationType(relfield, 'absolute')
        gotconvabs2 = nonlinear.convertDeformationType(absfield, 'relative')
    
        tol = dict(atol=1e-5, rtol=1e-5)
    
        assert np.all(np.isclose(gotconvrel1, absfield.data, **tol))
>       assert np.all(np.isclose(gotconvabs1, relfield.data, **tol))
E       AssertionError: assert False
E        +  where False = <function all at 0xb029a538>(array([[[[ True,  True,  True],\n         [False,  True,  True],\n         [ True,  True,  True],\n         ...,\n        ...True],\n         ...,\n         [ True,  True,  True],\n         [ True,  True,  True],\n         [ True,  True,  True]]]]))
E        +    where <function all at 0xb029a538> = np.all
E        +    and   array([[[[ True,  True,  True],\n         [False,  True,  True],\n         [ True,  True,  True],\n         ...,\n        ...True],\n         ...,\n         [ True,  True,  True],\n         [ True,  True,  True],\n         [ True,  True,  True]]]]) = <function isclose at 0xb0287580>(array([[[[ 4.99522028e+00,  3.42753495e+00,  1.51441400e-01],\n         [-2.80493113e-02, -2.30150966e+00, -4.19278070e...  [-4.55121509e-01, -2.88049823e+00, -1.23957147e+00],\n         [-1.32711055e+00, -4.51992160e+00,  3.62786100e+00]]]]), array([[[[ 4.99523363e+00,  3.42753495e+00,  1.51441400e-01],\n         [-2.80359599e-02, -2.30150966e+00, -4.19278070e...  [-4.55121509e-01, -2.88049823e+00, -1.23957147e+00],\n         [-1.32711055e+00, -4.51992160e+00,  3.62786100e+00]]]]), **{'atol': 1e-05, 'rtol': 1e-05})
E        +      where <function isclose at 0xb0287580> = np.isclose
E        +      and   array([[[[ 4.99523363e+00,  3.42753495e+00,  1.51441400e-01],\n         [-2.80359599e-02, -2.30150966e+00, -4.19278070e...  [-4.55121509e-01, -2.88049823e+00, -1.23957147e+00],\n         [-1.32711055e+00, -4.51992160e+00,  3.62786100e+00]]]]) = DeformationField(Numpy array, None).data
tests/test_transform/test_nonlinear.py:137: AssertionError
```

**Comment by pauldmccarthy** [comment]
Hmm, the first error seems to imply that the `tests/test_deprecated.py` file has been changed - the line numbers it is checking against are hard-coded in the file itself, so if the file has been modified, the test will fail. Do any of your build steps modify the files (e.g. changing hashbangs, stripping blank lines, etc)?

The second error is occurring in a test that is a bit numerically unstable - I'll bump up the tolerance for the next release (2.6, should be out in the next week or so).

**Comment by sanjayankur31** [comment]
> Hmm, the first error seems to imply that the `tests/test_deprecated.py` file has been changed - the line numbers it is checking against are hard-coded in the file itself, so if the file has been modified, the test will fail. Do any of your build steps modify the files (e.g. changing hashbangs, stripping blank lines, etc)?

Ah, yes---we remove unneeded shebangs before the build/install happens, so the first line is removed from `tests/test_deprecated.py` there.

The check phase where these tests are run only happens after the build and install phases. I could either:

- tweak the line numbers (reduce them by 1?)
- see if it is possible to omit this one file from shebang removal at the start and then do it after the tests are run.

**Comment by pauldmccarthy** [comment]
What I might do (because it's a bit of a silly thing to do) is change the test to determine the line numbers at run time (by reading the source file in at the time the test runs - still silly, but not quite as silly as the existing strategy).

**Comment by pauldmccarthy** [comment]
Both of these issues should have been resolved from about version 2.6.0 - feel free to re-open if you still have issues
